File: frontend/src/api/device.ts
import { http } from "@/utils/http";

export interface InfoEntity {
  id: number;
  inst_type: number;
  inner_ip_v_6: string;
  telecom_ip_v_6: string;
  unicom_ip_v_6: string;
  mobile_ip_v_6: string;
  inner_ip_v_4: string;
  telecom_ip_v_4: string;
  unicom_ip_v_4: string;
  mobile_ip_v_4: string;
  status: string;
  desc: string;
  group_id?: number;
}

export interface InfoRealTime {
  status_code: number;
  status: string;
  alloced: boolean;
  room_id: string;
  room_role_num: number;
}

export interface Device {
  id: number;
  name: string;
  deviceId: string;
  groupId: number;
  groupName: string;
  status: string;
  createdAt: string;
  updatedAt: string;
}

export interface GroupOption {
  id: number;
  name: string;
}

export interface DeviceListResult {
  code: number;
  message: string;
  data: {
    list: Device[];
    page: number;
    pageSize: number;
    total: number;
    groupOptions: GroupOption[];
  };
}

/** 获取设备列表 */
export function getDeviceList(params: { 
  page: number; 
  pageSize: number;
  keyword?: string;
  groupId?: number;
}) {
  return http.request<DeviceListResult>("get", "/api/devices/list", { params });
}

/** 保存设备 */
export function saveDevice(data: { 
  name: string;
  deviceId: string;
  groupId?: number;
  status: string;
}) {
  return http.request<{ code: number; message: string; data: Record<string, unknown> }>(
    "post",
    "/api/devices/create",
    { data }
  );
}

/** 设备截图响应结构 */
export interface ScreenshotResponse {
  code: number;
  message: string;
  data: {
    deviceId: string;
    success: boolean;
    imageData?: string; // base64编码的图片数据
    error?: string;     // 错误信息
  };
}

/** 后端直接返回的截图响应结构 */
interface ScreenshotBackendResponse {
  code: number;
  message: string;
  data: {
    deviceId: string;
    success: boolean;
    imageData?: string;
    error?: string;
  } | ScreenshotRes; // 可能是直接返回的结构
}

/** 后端可能直接返回的截图数据结构 */
interface ScreenshotRes {
  deviceId: string;
  success: boolean;
  imageData?: string;
  error?: string;
}

/** 获取设备截图 
 * @param data.deviceId 设备ID
 * @param data.quality 图片质量，1-100之间的整数，默认80
 * @returns 截图响应
 */
export function captureDeviceScreenshot(data: {
  deviceId: string;
  quality?: number;
}) {
  return http.request<ScreenshotBackendResponse>(
    "post",
    "/api/screenshot/capture",
    { data }
  ).then(response => {
    // 检查响应结构，进行适配转换
    if (response.code === 0) {
      // 检查data是否符合ScreenshotRes结构(直接返回的结构)
      const resData = response.data;
      if (!resData.deviceId && 'deviceId' in resData) {
        // 调整为标准化的响应格式
        return {
          code: response.code,
          message: response.message,
          data: response.data as ScreenshotRes
        };
      }
    }
    // 返回原始响应
    return response as unknown as ScreenshotResponse;
  });
}

/** 更新设备 */
export function updateDevice(data: { 
  id: number;
  name?: string;
  deviceId?: string;
  groupId?: number;
  status?: string;
}) {
  return http.request<{ code: number; message: string; data: Record<string, unknown> }>(
    "put",
    "/api/devices/update",
    { data }
  );
}

/** 删除设备 */
export function deleteDevice(id: number) {
  return http.request<{ code: number; message: string; data: Record<string, unknown> }>(
    "delete",
    "/api/devices/delete",
    { data: { id } }
  );
}

/** 批量回到主菜单响应结构 */
export interface BatchOperationResponse {
  code: number;
  message: string;
  data: {
    results: Record<string, string>;
  };
}

/** 批量回到主菜单 */
export function batchGoHome(deviceIds: string[]) {
  return http.request<BatchOperationResponse>(
    "post",
    "/api/devices/batch-go-home",
    { data: { deviceIds } }
  );
}

/** 批量清除后台应用 */
export function batchKillApps(deviceIds: string[]) {
  return http.request<BatchOperationResponse>(
    "post",
    "/api/devices/batch-kill-apps",
    { data: { deviceIds } }
  );
} 

File: backend/api/device/v1/device.go
package v1

import (
	"github.com/gogf/gf/v2/frame/g"
)

// DeviceStatus 设备状态枚举
const (
	DeviceStatusOnline  = "online"
	DeviceStatusOffline = "offline"
)

// 分组选项，用于前端下拉选择
type GroupOption struct {
	Id   int64  `json:"id"`
	Name string `json:"name"`
}

type ListReq struct {
	g.Meta   `path:"/devices/list" tags:"设备管理" method:"get" summary:"获取设备列表"`
	Page     int    `json:"page" v:"required#请输入页码" dc:"页码"`
	PageSize int    `json:"pageSize" v:"required#请输入每页数量" dc:"每页数量"`
	GroupId  int64  `json:"groupId" dc:"分组ID"`
	Keyword  string `json:"keyword" dc:"搜索关键词"`
}

type ListRes struct {
	List         []Device      `json:"list" dc:"设备列表"`
	Page         int           `json:"page" dc:"页码"`
	PageSize     int           `json:"pageSize" dc:"每页数量"`
	Total        int           `json:"total" dc:"总数"`
	GroupOptions []GroupOption `json:"groupOptions" dc:"分组选项列表"`
}

type CreateReq struct {
	g.Meta   `path:"/devices/create" tags:"设备管理" method:"post" summary:"创建设备"`
	Name     string `json:"name" v:"required#请输入设备名称" dc:"设备名称"`
	DeviceId string `json:"deviceId" v:"required|regex:^[a-zA-Z0-9_\\-\\:\\.]+$#请输入设备ID|设备ID只能包含字母、数字、下划线、中划线、冒号和点" dc:"设备ID，只能包含字母、数字、下划线、中划线、冒号和点"`
	GroupId  int64  `json:"groupId" dc:"分组ID，不指定则为0表示未分组"`
	Status   string `json:"status" v:"required|in:online,offline#请输入设备状态|设备状态只能是online或offline" dc:"设备状态：online-在线，offline-离线"`
}

type CreateRes struct{}

type UpdateReq struct {
	g.Meta   `path:"/devices/update" tags:"设备管理" method:"put" summary:"更新设备"`
	Id       int64  `json:"id" v:"required#请输入设备ID" dc:"设备ID"`
	Name     string `json:"name" v:"required#请输入设备名称" dc:"设备名称"`
	DeviceId string `json:"deviceId" v:"required|regex:^[a-zA-Z0-9_\\-\\:\\.]+$#请输入设备ID|设备ID只能包含字母、数字、下划线、中划线、冒号和点" dc:"设备ID，只能包含字母、数字、下划线、中划线、冒号和点"`
	GroupId  int64  `json:"groupId" dc:"分组ID，不指定则为0表示未分组"`
	Status   string `json:"status" v:"required|in:online,offline#请输入设备状态|设备状态只能是online或offline" dc:"设备状态：online-在线，offline-离线"`
}

type UpdateRes struct{}

type DeleteReq struct {
	g.Meta `path:"/devices/delete" tags:"设备管理" method:"delete" summary:"删除设备"`
	Id     int64 `json:"id" v:"required#请输入设备ID" dc:"设备ID"`
}

type DeleteRes struct{}

type Device struct {
	Id        int64  `json:"id"`
	Name      string `json:"name"`
	DeviceId  string `json:"deviceId"`
	GroupId   int64  `json:"groupId"`
	GroupName string `json:"groupName"`
	Status    string `json:"status"`
	CreatedAt string `json:"createdAt"`
	UpdatedAt string `json:"updatedAt"`
}

// BatchGoHomeReq 批量回到主菜单请求
type BatchGoHomeReq struct {
	g.Meta    `path:"/devices/batch-go-home" tags:"设备管理" method:"post" summary:"批量回到主菜单"`
	DeviceIds []string `json:"deviceIds" v:"required#请选择设备|length:1,50#设备数量必须在1-50之间" dc:"设备ID列表，最多50个"`
}

type BatchGoHomeRes struct {
	Results map[string]string `json:"results" dc:"操作结果，key为设备ID，value为错误信息（成功为空）"`
}

// BatchKillAppsReq 批量清除当前应用请求
type BatchKillAppsReq struct {
	g.Meta    `path:"/devices/batch-kill-apps" tags:"设备管理" method:"post" summary:"批量清除当前应用"`
	DeviceIds []string `json:"deviceIds" v:"required#请选择设备|length:1,50#设备数量必须在1-50之间" dc:"设备ID列表，最多50个"`
}

type BatchKillAppsRes struct {
	Results map[string]string `json:"results" dc:"操作结果，key为设备ID，value为错误信息（成功为空）"`
}
